lib_nn/api/geom/ImageGeometry.hpp

```diff
@@ -0,0 +1,236 @@
+   * An assertion is made that the specified location is within the geometry's
+   * bounds.
+   */
+  template <typename T>
+  T &Element(T *img_base, const int row, const int col,
+             const int channel) const;
+
+  /**
+   * Get the value of the specified element from the provided image.
+   * If the coordinates refer to a location outside the image geometry pad_value
+   * is returned.
+   */
+  template <typename T>
+  T Get(const T *img_base, const ImageVect coords, const T pad_value = 0) const;
+
+  /**
+   * Get the value of the specified element from the provided image.
+   * If the coordinates refer to a location outside the image geometry pad_value
+   * is returned.
+   */
+  template <typename T>
+  T Get(const T *img_base, const int row, const int col, const int channel,
+        const T pad_value = 0) const;
+
+  /**
+   * Apply an operation per element of an image.
+   *
+   * void PerPixelOp(const int row,
+   *                 const int col,
+   *                 const int channel,
+   *                 T& value);
+   */
+  template <typename T>
+  void ApplyOperation(T *image_base,
+                      std::function<void(int, int, int, T &)> func) const;
+
+  /**
+   * Determine whether this ImageGeometry is equal to another.
+   */
+  bool operator==(ImageGeometry other) const;
+};
+
+/////////////////////////
+template <typename T>
+T &ImageGeometry::Element(T *img_base, const int row, const int col,
+                          const int channel) const {
+  assert(IsWithinImage(row, col, channel));
+  int index = row * this->width * this->depth + col * this->depth + channel;
+  return img_base[index];
+}
+
+/////////////////////////
+template <typename T>
+T ImageGeometry::Get(const T *img_base, const ImageVect coords,
+                     const T pad_value) const {
+  return Get<T>(img_base, coords.row, coords.col, coords.channel, pad_value);
+}
+
+/////////////////////////
+template <typename T>
+T ImageGeometry::Get(const T *img_base, const int row, const int col,
+                     const int channel, const T pad_value) const {
+  if (!IsWithinImage(row, col, channel)) return pad_value;
+  return Element<T>(const_cast<T *>(img_base), row, col, channel);
+}
+
+/////////////////////////
+template <typename T>
+void ImageGeometry::ApplyOperation(
+    T *image_base, std::function<void(int, int, int, T &)> op) const {
+  for (int row = 0; row < this->height; ++row)
+    for (int col = 0; col < this->width; ++col)
+      for (int channel = 0; channel < this->depth; ++channel)
+        op(row, col, channel, this->Element<T>(image_base, row, col, channel));
+}
+
+/////////////////////////
+inline std::ostream &operator<<(std::ostream &stream,
+                                const ImageGeometry &image) {
+  return stream << image.height << "," << image.width << "," << image.depth
+                << "," << image.channel_depth;
+}
+
+}  // namespace nn
```

Comment:
Could you please the files to have an extra line at the end.

Reply:
yeah, no, problem. Apparently clang-format doesn't do this for me!